Umbraco CI/CD Flow - first take (#5476)
umbraco/UmbracoDocs: umbraco-cloud/set-up/project-settings/umbraco-cicd/SampleCiCdPipeline.md

# Getting started with configuring a CI/CD pipeline

Learn how to configure a CI/CD pipeline in Azure DevOps and GitHub Actions Workflows using the scripts provided.

You'll find example shell scripts and pipeline configurations in the section "Sample scripts", covering both Azure DevOps and GitHub Actions Workflows.

# Why should one configure a sample CI/CD pipeline?
"Umbraco Cloud repositories are not meant to be used as source code repositories. More details on our official documentation.
Once you commit your code to Cloud the build pipeline converts your C# code to DLLs and deploys it on the respective environment.
It is important to note that only C# code is built, and all frontend artifacts need to be built and committed to the repository.
You can use AzureDevops as an external repository and with the pipelines it will automatically keep your azure devops source code repository in sync with the git repository of Umbraco Cloud of development environment.

> !INSERT IMAGE AS PROVIDED BY ALINA!

Reviewer: Remember to insert the image that Alina have provided :)

Author: This has now been added.

# Getting started with configuring a CI/CD pipeline

Learn how to configure a CI/CD pipeline in Azure DevOps and GitHub Actions Workflows using the scripts provided.

You'll find example shell scripts and pipeline configurations in the "Sample scripts" section, covering both Azure DevOps and GitHub Actions Workflows.

# Why should one configure a sample CI/CD pipeline?
Umbraco Cloud repositories are not meant to be used as source code repositories. More details on our official documentation.

Once you commit your code to Cloud the build pipeline converts your C# code to DLLs and deploys it on the respective environment.

> It is important to note that only C# code is built, and all frontend artifacts need to be built and committed to the repository.

Reviewer 2: ```suggestion
{% hint style="info" %}

Only C# code is built, and all frontend artifacts need to be built and committed to the repository.

{% endhint %}

```

> You can use AzureDevops as an external repository and with the pipelines, it will automatically keep your Azure Devops source code repository in sync with the git repository of Umbraco Cloud of the development environment.

Reviewer 2: ```suggestion
You can use AzureDevops as an external repository and with the pipelines, it will automatically keep your Azure Devops source code repository in sync. The sync is done with the git repository of Umbraco Cloud of the development environment.
```

> ![UmbracoCloud CICD sample pipeline](../../images/UmbracoCloudCicdSample.png)

Reviewer 2: ```suggestion
![UmbracoCloud CI/CD sample pipeline](../../images/UmbracoCloudCicdSample.png)
```

# Prerequisites: Setting Up an Umbraco Cloud Project

> Before proceeding, you'll need an Umbraco Cloud project, a CI/CD pipeline, and the required files to add to your pipeline for successful interaction with the Umbraco Cloud API.

Reviewer 2: ```suggestion
Before proceeding, you'll need an Umbraco Cloud project and a CI/CD pipeline. You will also need the required files to add to your pipeline for successful interaction with the Umbraco Cloud API.
```

1. Create an Umbraco Cloud Project: Preferably with a development environment. You can either create a trial,  create a new project, or use an existing one.
2. Create a new or use an existing CI/CD pipeline: For Azure DevOps see this [page](https://learn.microsoft.com/en-us/azure/devops/organizations/projects/create-project?view=azure-devops&tabs=browser). 
3. Get your set of supporting files: Download sample files as a [zip-file](https://drive.google.com/file/d/1Wpxib2F5-eIyVsSdm3EBvA54eOrgiwLi/view?usp=drive_link). Note, that the only file you properly need to make changes to is the .yaml file. 

> **Note**: On the page "What is Umbraco CI/CD Flow", deployments are targeted at the leftmost environment in your Umbraco Cloud setup. This means if you have a Development environment, it will be automatically selected for deployment. If no Development environment exists, the Live environment will be used.

Reviewer 2: ```suggestion
{% hint style="info" %}

On the page "What is Umbraco CI/CD Flow", deployments are targeted at the leftmost environment in your Umbraco Cloud setup. This means if you have a Development environment, it will be automatically selected for deployment. If no Development environment exists, the Live environment will be used.

{% endhint %}

```

## Obtaining the Project ID and API Key

To get started with API interactions, you'll need to obtain your Project ID and API key. If you haven't already enabled the CI/CD feature, follow these steps:

1. Navigate to the [Umbraco Cloud Portal](https://www.s1.umbraco.io/projects) and select your project.
2. Go to `Settings` -> `Advanced`. This is where you can generate an API key and find your Project ID.

> **Important Note**: The API key is tied to the specific project for which it is generated. Make sure to keep it secure, as it will be used for all subsequent API interactions related to that project.

Reviewer 2: ```suggestion

{% hint style="info" %}

The API key is tied to the specific project for which it is generated. Make sure to keep it secure, as it will be used for all subsequent API interactions related to that project.

{% endhint %}

```

## Cloning the Umbraco Cloud Repository

> To get started with your local development, you'll need to clone the repository of the environment you intend to work with—either the development or live environment. Once cloned, make a copy of it in a new folder. In this example, we'll create a new folder named `local-cicd-demo-site` for the local repository.

Reviewer 2: ```suggestion
To get started with your local development, you'll need to clone the repository of the environment you intend to work with. Once cloned, make a copy of it in a new folder. In this example, we'll create a new folder named `local-cicd-demo-site` for the local repository.

```

2. **Configure Variables**: Open the `azure-release-pipeline.yaml` file and set the appropriate variables in the variables section. These variables can be found in the Settings -> Advanced section of your project on the [Umbraco Cloud Portal](https://www.s1.umbraco.io/projects).

3. **Variable Requirements**: The sample pipeline expects the following variables to be set:
    * `umbraco-cloud-api-key`: The API key needed to access the Umbraco Cloud API.
    * `git-pat`: A Personal Access Token (PAT) for Git interactions.
    * `project-id`: The ID of the project that the pipeline will be used for.

By following these steps, you'll have a staged pipeline configured in Azure DevOps tailored for your Umbraco Cloud project.

### Pipeline Stage: Preflight Checks

This stage involves making an API call to retrieve the latest completed deployment. The goal is to identify any remote changes that may have occurred since the last successful deployment. If such changes are detected, they are applied to a new Git branch and pushed. A manual validation gate is introduced if changes are found.

> Note: The current pipeline script has room for improvement. For instance, it may falsely indicate remote changes even after a patch has been applied. It also requires manual intervention twice, as changes persist until they are deployed.

Reviewer 2: ```suggestion
{% hint style="info" %}

The current pipeline script has room for improvement. For instance, it may falsely indicate remote changes even after a patch has been applied. It also requires manual intervention twice, as changes persist until they are deployed.

{% endhint %}

```

# Getting started with configuring a CI/CD pipeline

> Learn how to configure a CI/CD pipeline in Azure DevOps and GitHub Actions Workflows using the scripts provided.

Reviewer 2: ```suggestion

{% hint style="warning" %}
This documentation is currently a work in progress.
{% endhint %}

Learn how to configure a CI/CD pipeline in Azure DevOps and GitHub Actions Workflows using the scripts provided.
```

```sh
# Clone the development environment repository
git clone https://scm.umbraco.io/euwest01/dev-cicd-demo-site.git

# Copy the repository to a new local folder
cp -r dev-cicd-demo-site local-cicd-demo-site
```

This will set up your local workspace, allowing you to work on the project before pushing changes back to the Umbraco Cloud repository.

## Reconfiguring Git Remotes

After cloning the Umbraco Cloud repository, it's essential to remove its remote settings so that you can link it to your own company's repository. In this example, we'll be using an Azure DevOps-hosted repository as the new origin. Follow the steps below to reset the Git remote from the root folder of `local-cicd-demo-site`:

```sh
# Reset the Git remote to point to the new Azure DevOps repository
git remote set-url origin git@ssh.dev.azure.com:v3/umbraco/Cloud%20Team/local-cicd-demo-site
```

By executing this command, you'll disconnect the local repository from the Umbraco Cloud repository and connect it to your company's Azure DevOps repository. This allows you to manage your project within your own version control system.

## Optional: Renaming the Master Branch to Main

> If you prefer to use the term "main" instead of "master" for your primary branch, you can easily rename it with the following commands:

Reviewer 2: ```suggestion
If you prefer to use the term "main" instead of "master" for your primary branch, you can rename it with the following commands:
```

```sh
# Rename the 'master' branch to 'main'
git branch -m master main

# Push the newly renamed 'main' branch to the remote repository
git push -u origin main
```

By executing these commands, you'll rename the local 'master' branch to 'main' and update the remote repository to reflect this change. This is an optional step but aligns with the industry trend towards more inclusive language.

## Setting Up the Pipeline in Azure DevOps

> This section provides a step-by-step guide to setting up a CI/CD pipeline in Azure DevOps using the provided sample scripts[link to "Sample scripts" section of this document]. The pipeline is defined in an Azure YAML file and includes several steps that call custom shell scripts to interact with the Umbraco Cloud API.

Reviewer 2: ```suggestion
This section provides a step-by-step guide to setting up a CI/CD pipeline in Azure DevOps using the provided sample scripts[link to "Sample scripts" section of this document]. The pipeline is defined in an Azure YAML file and includes some steps that call custom shell scripts to interact with the Umbraco Cloud API.
```

3. **Variable Requirements**: The sample pipeline expects the following variables to be set:
    * `umbraco-cloud-api-key`: The API key needed to access the Umbraco Cloud API.
    * `git-pat`: A Personal Access Token (PAT) for Git interactions.
    * `project-id`: The ID of the project that the pipeline will be used for.

By following these steps, you'll have a staged pipeline configured in Azure DevOps tailored for your Umbraco Cloud project.

### Pipeline Stage: Preflight Checks

This stage involves making an API call to retrieve the latest completed deployment. The goal is to identify any remote changes that may have occurred since the last successful deployment. If such changes are detected, they are applied to a new Git branch and pushed. A manual validation gate is introduced if changes are found.

{% hint style="info" %}

The current pipeline script has room for improvement. For instance, it may falsely indicate remote changes even after a patch has been applied. It also requires manual intervention twice, as changes persist until they are deployed.

{% endhint %}

### Pipeline Stage: Build, Test, and Package

> In this stage, the Umbraco CMS project is built and tested. While this step is optional, it provides significant value, especially if you're running custom frontend builds with different npm versions for both the back office and website views.

Reviewer 2: ```suggestion
In this stage, the Umbraco CMS project is built and tested. While this step is optional, it provides significant value. Especially if you're running custom frontend builds with different npm versions for both the backoffice and website views.
```
